### PR TITLE
Map CRDT + Op based replication + CvRDT/CmRDT/Causal Traits + More Tests + Many other changes

### PR DESCRIPTION
**Pull request by davidrusu**
Sorry about the mega PR. This changeset is the accumulation of all the work done in #5, #6 and now also Op based replication.

A summary of the changes:
- **traits** Introduce`trait Causal<A: Actor>, trait CvRDT, trait CmRDT`
- **map** Add's a Map CRDT which uses the same mechanism as ORSWOT for resolving conflicts.
  - The map supports composing CRDT's with the same reset-remove semantics from riak's map
- **trait alias'** Create some trait alias' like the `vclock::Actor` trait and use them across them consistently across the library
- **api changes** Changes to the `VClock` and `LWWReg` api
  - most important one being `VClock::get(&actor)` will return 0 if the actor does not have a dot.
- **Op based replication** CRDT's that implement the `CmRDT` trait are ob 
  - I have a well tested op based replication implementation for the Map CRDT
  - The ORSWOT has a bare bones implementation but it needs further testing (I'd like to do that work in another PR)
  - LWWReg is small enough that we can ship the entire state as an op and be done with it, this is the current implementation.

@spacejam

### CODE REVIEW

File: src/vclock.rs [hunk 1]
```diff
@@ -10,17 +10,21 @@
 //! assert!(a > b);
 //! ```
 
+// TODO: we have a mixture of language here with witness and actor. Clean this up
+
```
Comment (spacejam) [inline src/vclock.rs]: Thinking about what might be less confusing for newcomers, I think `witness` might be more clear for them. What do you think about changing `Actor` to `Witness` to try to minimize confusion?

**Comment by spacejam** [comment]
Sorry it's taken so long for me to look at this work. I have a bit more free time now, and am excited to look at this more today :]

File: src/map.rs [hunk 2]
```diff
@@ -745,6 +745,37 @@ mod tests {
...
+
+        m3.apply(&m2_rm);
+        m3.apply(&m1_up1);
+        m3.apply(&m1_up2);
+
+        m1.apply(&m2_rm);
+
+        assert_eq!(m2, m3);
+        assert_eq!(m1, m2);
+        assert_eq!(m1, m3);
```
Comment (davidrusu) [inline src/map.rs]: I thought I could simplify the implementation of the map compared to what riak had... I was thinking last night and I realized this implementation breaks down in this test case. 

I need to look into why quickcheck didn't catch this case, I'll see about improving  those tests.
@spacejam

File: src/vclock.rs (hunk 1 above)
Comment (davidrusu) [reply src/vclock.rs]: I like `witness`, I'll make the change :)

**Comment by davidrusu** [comment]
there's a few areas that need improvements but lets do that in future PR's, this is large enough already

**Pull request merged**